cupy/cuda/function.pyx

```diff
@@ -90,6 +90,11 @@ cdef class CIntptr(CPointer):
         self.ptr = <void*>&self.val
 
 
+cdef class CNumpyArray(CPointer):
+    def __init__(self, a):
+        self.ptr = <void*><size_t>a.__array_interface__['data'][0]
```

Comment:
I think it's better to hold the reference to the numpy array, in case it goes out of scope (say we pass in a temporary numpy array). Something like
```suggestion
cdef class CNumpyArray(CPointer):
    cdef:
        object val

    def __init__(self, a):
        self.val = a
        self.ptr = <void*><intptr_t>a.ctypes.data
```

Reply:
Also, accessing `a.ctypes.data` should be slightly faster.

Reply:
(Tests are running, patch will be pushed soon)

1. I fixed the signature. The type error was however already raised before this change with the message (this is checked in `tests/cupy_tests/core_tests/test_function.py:106`). Removing the context manager will trigger the new type error.
2. Hold a reference to the ndarray: done
3. About the performance, `__array_interface__` seems to do slightly better than `ctypes`:

**Array interface:**
```
python3 -m timeit -s 'import numpy as np; a=np.empty((8,8), dtype=np.float32);' -n 100000 -- 'a.__array_interface__["data"][0]'
100000 loops, best of 3: 0.738 usec per loop
```

**Ctypes interface**
```
python3 -m timeit -s 'import numpy as np; a=np.empty((8,8), dtype=np.float32);' -n 100000 -- 'a.ctypes.data'
100000 loops, best of 3: 0.846 usec per loop
```

The only possible improvement would be to use directly `numpy.ndarray.data` (0.165 usec) which returns a memory view. The problem is that I did not manage to obtain a pointer from this python memory view without first declaring a cython memory view with cdef. I did not manage to cdef a generic memory view object (both in dimension and in type) so my initial solution was to:
- viewing the array as a one dimensional array with `numpy.ndarray.ravel`
- viewing it as a "compile time" known type (like `np.uint8`) with `numpy.ndarray.view`
as it can be seen in my first commit.

Reply:
> 1. The signature cannot be fixed with your change, cython build fails with
>    `Exception clause not allowed for function returning Python object`.

Thank you, @keckj. Apology, I shouldn't have rushed into comments right before leaving from work. `CPointer` is a `cdef class`, which is indeed a Python object, so there's no problem in propagating any exception.

Re: `__array_interface__`, thanks for testing it again. I just recalled that you already tested it earlier (https://github.com/cupy/cupy/pull/4829#discussion_r589392268) 😅 I think it's fine to leave it as is. Any other option is more complex than this or `a.ctypes.data`.

Reply:
> According to my very limited cython knowledge you cannot cdef a generic memory view object (both in dimension and in type)

I think you are right on this. AFAIK Cython's memoryview is only limited to primitive (C) types, so is Python's built-in `array` module.